### PR TITLE
[SFTTrainer]: Compute accuracy in log to avoid gather every step.

### PR DESCRIPTION
**Pull request by pramodith**
# What does this PR do?

Computes the accuracy in the `log` function to avoid `gather` calls after every forward step.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by qgallouedec** [comment]
So ideally the log should just _log_ (and not compute metrics), but if it really helps improving the speed, maybe we can make an exception here. Do you have a benchmark, to see it speeds things up?

**Comment by pramodith** [comment]
> So ideally the log should just _log_ (and not compute metrics), but if it really helps improving the speed, maybe we can make an exception here. Do you have a benchmark, to see it speeds things up?

I ran this code block with and without the changes and only got a 0.3 second improvement (might be noise) so I guess its not worth it?

GPUs -> 2 A10s
<details>
<summary>Code</summary>

```
from trl import SFTTrainer, SFTConfig
from peft import LoraConfig
from datasets import load_dataset
import time
from accelerate import PartialState
from transformers import AutoModelForCausalLM


device_map="DDP" # for DDP and running with 

if device_map == "DDP":
    device_string = PartialState().process_index
    device_map={'':device_string}

model = AutoModelForCausalLM.from_pretrained(
    "Qwen/Qwen3-0.6B",
    trust_remote_code=True,
    use_cache = False,
    device_map = device_map,  # Use DDP device map
)

config = SFTConfig(
    per_device_train_batch_size=4,
    max_steps=100,
    logging_steps=10,
    gradient_checkpointing=True,
    gradient_checkpointing_kwargs = {"use_reentrant": False}, #must be false for DDP
)


trainer = SFTTrainer(
    model=model,
    args=config,
    train_dataset=load_dataset("trl-lib/Capybara", split="train"),
    peft_config=LoraConfig(),
    
)

start_time = time.time()
trainer.train()
print(f"Training took {time.time() - start_time} seconds")
```
</details>

**Comment by qgallouedec** [comment]
I'll try with 2 nodes of 8 H100

**Comment by qgallouedec** [comment]
After extensive benchmarking, I found that gathering introduces negligible overhead. With your code and 2 nodes of 8×H100, the gathering time is only about 0.9 second / 673 seconds (0.13% of training time). Plus, here, we’re training a small model, this is a worst-case scenario.

**Comment by pramodith** [comment]
> After extensive benchmarking, I found that gathering introduces negligible overhead. With your code and 2 nodes of 8×H100, the gathering time is only about 0.9 second / 673 seconds (0.13% of training time). Plus, here, we’re training a small model, this is a worst-case scenario.

Thanks for running the benchmark! This is an interesting and good learning for me, didn't expect the gather operations to be that efficient.
